Synopsis/Parsers/Cxx/ucpp/ucpp.1
.TH UCPP 1 "Oct 21 2000"
.SH NAME
ucpp \- C preprocessor
.SH SYNOPSIS
.B ucpp 
[
.I options
]
[
.I file
]
.SH DESCRIPTION
.LP
.B ucpp
is a C preprocessor mostly compatible with ISO-C99.
It is rather strict and uses only a small amount of memory. It uses
standard input as primary input if no file argument is given.
.SH OPTIONS
There are several classes of options.
.TP
.B Language Options
.TP
.BI \-C
keep comments in the output.
.TP
.BI \-s
if a rogue '#' is encountered, do not emit an error and keep it in
the output.
.TP
.BI \-l
supress the emission of '#line' directives in the output.
.TP
.BI \-lg
convert the '#line' to the gcc-style equivalent.
.TP
.BI \-CC
disable C++-like comments (a '//' begins a comment, up to the end
of the line). Use this option to get closer to C90 behaviour.
.TP
.B \-a, \-na
handle assertions (defined with #assert);
.B \-a
also defines the standard assertions
.I #machine
,
.I #cpu
and
.I #system
(see
.B \-e
to get the local definition of such assertions).
.TP
.BI \-a0
disable assertion support.
.TP
.BI \-V
disable support for macros with a variable number of arguments: in C99,
a macro may be declared with
.I ...
as the last argument; inside the replacement list,
.I __VA_ARGS__
is replaced with the optional extra arguments given in the call to the macro.
Use this option to get closer to C90 behaviour.
.TP
.BI \-u
enable UTF-8 support: with this option, the source is considered as
an ISO/10646 source, encoded in UTF-8. Characters represented as two bytes
or more are considered as alphabetic characters, like letters, and
therefore usable in identifiers. These characters hold the same
syntactic value than the corresponding Universal Character Names.
.TP
.BI \-X
enable
.B \-a, \-u
and
.B \-Y.
This should make
.B ucpp
behave closer to what is requested from a "modern" C preprocessor.
.TP
.BI \-c90
enable
.B \-V
and
.B \-CC,
and do not define
.B __STDC_VERSION__.
This should make
.B ucpp
mimic older C90 behaviour.
.TP
.BI \-t
disable trigraph support; this seems to be required for some legacy code.
.TP
.B Warning Options
.TP
.BI \-wt
emit a final warning when trigraphs are encountered.
.TP
.BI \-wtt
emit warnings for each trigraph encountered.
.TP
.BI \-wa
emit annoying warnings (these are usually useless).
.TP
.BI \-w0
supress standard warnings.
.TP
.B Directory Options
.TP
.BI \-I directory
.TP
.BI "\-I " directory
add
.I directory
to the include path, before the standard include path.
.TP
.BI "\-J " directory
add
.I directory
to the include path, after the standard include path.
.TP
.BI \-zI
do not use the standard (compile-time) include path.
.TP
.BI \-M
emit only the names of encountered files, separated by spaces; this is
intended for automatic generation of Makefile dependencies.
.TP
.BI \-Ma
do the same as
.B \-M
but also for system files.
.TP
.BI "\-o " file
direct the ouput to
.I file
instead of standard output.
.TP
.B Macro Options
.TP
.BI \-D macro
predefine
.I macro
with content
.B 1.
.TP
.BI \-D macro=def
predefine
.I macro
with the content
.I def.
.TP
.BI \-U macro
undefine
.I macro.
.TP
.BI \-Y
predefine system-dependant macros.
.TP
.BI \-Z
do not predefine special macros such as
.B __TIME__.
.TP
.BI \-A foo(bar)
add
.I foo(bar)
to the list of assertions.
.TP
.BI \-B foo(bar)
remove
.I foo(bar)
of the list of assertions; you may also use
.BI \-B foo
to remove all
.BI \-B foo(xxx)
from the list of assertions.
.TP
.BI \-d
instead of normal output, emit '#define' directives representing all
macros defined during processing.
.TP
.BI \-e
instead of normal output, emit '#assert' directives representing all
assertions defined during processing.
.TP
.B Miscellaneous Options
.TP
.BI \-v
print version number, include path and (optionaly) defined assertions.
.TP
.BI \-h
print some help.
.SH ENVIRONMENT
.PP
.B ucpp
is not itself affected by environment variables. However, it uses
library functions that might be affected, depending on the system.
.SH AUTHOR
Thomas Pornin <thomas.pornin@ens.fr>
.SH BUGS
.PP
.B ucpp
is considered stable software. However improbable it is, please report
bugs to the author (possibly with a file that exhibits the problem) if
the latest version, available from this site:
.TP
http://www.di.ens.fr/~pornin/ucpp/
.PP
has the bug.


NEWS
NEWS
====

This file describes user-visible changes to Synopsis.

Synopsis 0.12
=============

* Upgraded preprocessor to ucpp 1.3.
* Enhanced Cpp processor to allow documenting macros, as well as cross-referencing
  macro calls.
* Improvements for Python 2.6 / 3.x compatibility.

Synopsis 0.11
=============

* Enhance the build system to support OSX.
* Upgrade to libgc 7.0.
* Add a new '--disable-threads' config option.
* Fix the way template specializations are handled in the ASG.
* Fix a number of minor incompatibilities with Python 2.6.
* Fix a number of validation errors for the HTML formatter.

Synopsis 0.10
=============

Synopsis' Internal Representation has been largely refactored, to better
support the new features:

* More information is captured, for example relationships between primary
  class templates and their specializations / instantiations.

* The storage of source-code cross-references has been simplified and integrated
  into the IR, leading to a much more intuitive and convenient user interface.

The Python parser has been largely rewritten, adding support
for parsing of dependent (imported) files.

The DocBook formatter has been rewritten. It now supports Javadoc and ReST
document markup.

The HTML formatter has been enhanced in many respects. Many bugs have been fixed.

The C and C++ parsers have been enhanced, and many bugs have been fixed.

Some new ASG processors have been added (MacroFilter, TypedefFolder, et al.)

Documentation has been enhanced.


Synopsis 0.9
============

Support for C99 parsing has been greatly enhanced.

Support for Python code cross-referencing has been added.

The processing of source-embedded documentation has been greatly
enhanced. As a result, Synopsis now supports multiple markup formats,
such as Javadoc, and ReStructuredText. The latter requires python-docutils
to be installed.

Synopsis no longer depends on an external omniidl tool, as its IDL
parser is now self-contained.

Solidified and optimized SXR server.

Much enhanced HTML formatter.

Etc.


Synopsis/Parsers/Cpp/ucpp/ucpp.cc
//
// Copyright (C) 2003 Stefan Seefeld
// All rights reserved.
// Licensed to the public under the terms of the GNU LGPL (>= 2),
// see the file COPYING for details.
//

#include <Synopsis/AST/ASTKit.hh>
#include <Synopsis/AST/TypeKit.hh>
#include <Support/Path.hh>
#include <Support/ErrorHandler.hh>
#include <vector>
#include <string>
#include <iostream>
#include <fstream>
#include <cstring>
#include <cstdio>
#include <memory>
#include <functional>

using namespace Synopsis;

namespace
{

int verbose = 0;
int debug = 0;
int main_file_only = 0;
bool active = true;
const char *language;
std::string base_path;
// these python objects need to be referenced as pointers
// since the python runtime environment has to be initiaized first
std::auto_ptr<AST::ASTKit> kit;
std::auto_ptr<AST::TypeKit> types;
std::auto_ptr<AST::AST> ast;
std::auto_ptr<AST::SourceFile> source_file;
const char *input = 0;

//. creates new SourceFile object and store it into ast
AST::SourceFile create_source_file(const std::string &filename, bool is_main)
{
  Path path = Path(filename).abs();
  path.strip(base_path);
  std::string name = path.str();
  AST::SourceFile sf = kit->create_source_file(name, filename);
  Python::Dict files = ast->files();
  files.set(name, sf);
  if (is_main) sf.is_main(true);
  return sf;
}

//. creates new or returns existing SourceFile object
//. with the given filename
AST::SourceFile lookup_source_file(const std::string &filename, bool main)
{
  Python::Dict files = ast->files();
  Path path = Path(filename).abs();
  path.strip(base_path);
  AST::SourceFile sf = files.get(path.str());
  if (sf && main) sf.is_main(true);
  return sf ? sf : create_source_file(filename, main);
}

//. creates new Macro object
void create_macro(const char *filename, int line,
		  const char *macro_name, int num_args, const char **args,
		  int vaarg, const char *text)
{
  Python::Tuple name(macro_name);
  AST::SourceFile sf = lookup_source_file(filename, true);
  Python::List params;
  if (args)
  {
    for (int i = 0; i < num_args; ++i) params.append(args[i]);
    if (vaarg) params.append("...");
  }
  AST::Macro macro = kit->create_macro(sf, line, name, params, text);
  AST::Declared declared = types->create_declared(name, macro);

  Python::List declarations = ast->declarations();
  declarations.append(macro);

  // FIXME: the 'types' attribute is not (yet) a dict type
  // so we have to do the call conversions manually...
  Python::Object types = ast->types();
  types.attr("__setitem__")(Python::Tuple(name, declared));
}

bool extract(PyObject *py_flags, std::vector<const char *> &out)
{
  Py_INCREF(py_flags);
  Python::List list = Python::Object(py_flags);
  for (size_t i = 0; i != list.size(); ++i)
  {
    const char *value = Python::Object::narrow<const char *>(list.get(i));
    if (!value) return false;
    out.push_back(value);
  }
  return true;
}

extern "C"
{
  // ucpp_main is the renamed main() func of ucpp, since it is included in this
  // module
  int ucpp_main(int argc, char** argv);
}

PyObject *ucpp_parse(PyObject *self, PyObject *args)
{
  try
  {
    const char *py_base_path;
    char *output;
    PyObject *py_flags;
    std::vector<const char *> flags;
    PyObject *py_ast;
    if (!PyArg_ParseTuple(args, "OszzsO!iii",
                          &py_ast,
                          &input,
                          &py_base_path,
                          &output,
                          &language,
                          &PyList_Type, &py_flags,
                          &main_file_only,
                          &verbose,
                          &debug)
        || !extract(py_flags, flags))
      return 0;
    
    Py_INCREF(py_ast);
    // since everything in this file is accessed only during the execution
    // of ucpp_parse, we can safely manage these objects in this scope yet
    // reference them globally (for convenience)
    ast.reset(new AST::AST(py_ast));
    kit.reset(new AST::ASTKit(language));
    types.reset(new AST::TypeKit(language));

    if (py_base_path)
    {
      Path path(py_base_path);
      base_path = path.abs().str();
    }

    source_file.reset(new AST::SourceFile(lookup_source_file(input, true)));

    flags.insert(flags.begin(), "ucpp");
    flags.push_back("-C"); // keep comments
    flags.push_back("-lg"); // gcc-like line numbers
    if (output)
    {
      flags.push_back("-o"); // output to...
      flags.push_back(output);
    }
    else
    {
      flags.push_back("-o"); // output to...
#ifdef __WIN32__
      flags.push_back("NUL");
#else
      flags.push_back("/dev/null");
#endif
    }
    flags.push_back(input);
    if (verbose)
    {
      std::cout << "calling ucpp\n";
      for (std::vector<const char *>::iterator i = flags.begin();
	   i != flags.end(); ++i)
	std::cout << ' ' << *i;
      std::cout << std::endl;
    }
    
    int status = ucpp_main(flags.size(), (char **)&*flags.begin());
    if (status != 0)
      std::cerr << "ucpp returned error flag. ignoring error." << std::endl;

    Python::Dict files = ast->files();
    Path path = Path(input).abs();
    path.strip(base_path);
    files.set(path.str(), *source_file);

    py_ast = ast->ref(); // add new reference
    // make sure these objects are deleted before the python runtime
    source_file.reset();
    types.reset();
    kit.reset();
    ast.reset();
    return py_ast;
  }
  catch (const std::exception &e)
  {
    std::cerr << "Internal error : " << e.what() << std::endl;
    return 0;
  }
}

extern "C"
{
  //. This function is a callback from the ucpp code to report
  //. a context switch in the parser, i.e. either to enter a
  //. new (included) file (@new_file == true) or to return control
  //. to an already open file once parsing an included file is
  //. terminated (@new_file == false)
  //. Depending on whether the new filename matches with the base_path,
  //. the parser is activated or deactivated
  void synopsis_file_hook(const char *filename, int new_file)
  {
    // turn 'filename' into an absolute path so we can match it against
    // base_path
    std::string abs_filename = Path(filename).abs().str();

    bool activate = false;
    if ((main_file_only && strcmp(input, filename)) || 
	(base_path.size() && abs_filename.substr(0, base_path.size()) != base_path))
      active = false;
    else
    {
      if (!active) active = activate = true;
    }

    if (!active) return;

    // just for symmetry, don't report if we were just activated 
    if (debug && !activate)
      if (new_file)
	std::cout << "entering new file " << abs_filename << std::endl;
      else
	std::cout << "returning to file " << abs_filename << std::endl;

    source_file.reset(new AST::SourceFile(lookup_source_file(abs_filename, true)));
  }

  //. This function is a callback from the ucpp code to store macro
  //. expansions
  void synopsis_macro_hook(const char *name, int line_num,
			   int start, int end, int diff)
  {
    if (!active) return;
    if (debug) 
      std::cout << "macro : " << name << ' ' << line_num << ' '
		<< start << ' ' << end << ' ' << diff << std::endl;

    Python::Dict mmap = source_file->macro_calls();
    Python::List line = mmap.get(line_num, Python::List());
    line.append(kit->create_macro_call(name, start, end, diff));
    mmap.set(line_num, line);
  }

  //. This function is a callback from the ucpp code to store includes
  void synopsis_include_hook(const char *source, const char *target,
			     const char *fname, int is_system,
			     int is_macro, int is_next)
  {
    if (!active) return;

    std::string name = fname;
    if (is_system) name = '"' + name + '"';
    else name = '<' + name + '>';

    if (debug) 
      std::cout << "include : " << source << ' ' << target << ' ' << name << ' '
		<< is_macro << ' ' << is_next << std::endl;

    Path path = Path(target).abs();

//     bool main = !base_path || strncmp(path.str().c_str(), base_path, strlen(base_path));
    AST::SourceFile target_file = lookup_source_file(path.str(), false);

    AST::Include include = kit->create_include(target_file, name, is_macro, is_next);
    Python::List includes = source_file->includes();
    includes.append(include);
  }

  //. This function is a callback from the ucpp code to store macro
  //. definitions
  void synopsis_define_hook(const char *filename, int line,
			    const char *name,
			    int num_args, const char **args, int vaarg, 
			    const char *text)
  {
    if (!active) return;
    if (debug) 
      std::cout << "define : " << filename << ' ' << line << ' ' 
		<< name << ' ' << num_args << ' ' << text << std::endl;

    create_macro(filename, line, name, num_args, args, vaarg, text);
  }
};

PyMethodDef methods[] = {{(char*)"parse", ucpp_parse, METH_VARARGS},
			 {0, 0}};
};

extern "C" void initucpp()
{
  Python::Module module = Python::Module::define("ucpp", methods);
  module.set_attr("version", "0.1");
}
